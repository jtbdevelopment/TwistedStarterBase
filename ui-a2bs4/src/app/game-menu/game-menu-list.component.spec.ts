import {fakeAsync, inject, TestBed, tick} from '@angular/core/testing';
import {BehaviorSubject, Observable} from 'rxjs';
import {Component, Input} from '@angular/core';
import {GameMenuListComponent} from './game-menu-list.component';
import {GameClassifier} from '../core-games-ui/gamecache/game-classifier.serviceinterface';
import {HelpDisplayService} from '../help/help-display.service';
import {NgbModule, NgbPopoverConfig} from '@ng-bootstrap/ng-bootstrap';

class MockGameClassifier {
    public static classifications: BehaviorSubject<string[]> = new BehaviorSubject<string[]>([]);
    public static icons: BehaviorSubject<Map<string, string>> = new BehaviorSubject<Map<string, string>>(new Map<string, string>());

    public getClassifications(): Observable<string[]> {
        return Observable.from(MockGameClassifier.classifications);
    }

    //  Game buckets to icons - can return empty map if not initialized
    public getIcons(): Observable<Map<string, string>> {
        return Observable.from(MockGameClassifier.icons);
    }

}

@Component({
    selector: 'game-menu-category-list',
    template: '<p>{{category}}</p><p>{{style}}</p><p>{{icon}}</p>'
})
class MockCategoryListComponent {
    @Input() public category: string;
    @Input() public icon: string;
    @Input() public style: string;
}

describe('Component:  game menu list component', () => {
    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                NgbModule
            ],
            declarations: [
                GameMenuListComponent,
                MockCategoryListComponent
            ],
            providers: [
                {provide: 'GameClassifier', useClass: MockGameClassifier},
                HelpDisplayService,
                NgbPopoverConfig
            ]
        });
        TestBed.compileComponents();
    });

    it('initializes', () => {
        const fixture = TestBed.createComponent(GameMenuListComponent);

        fixture.detectChanges();
        expect(fixture.componentInstance.categories).toEqual([]);
        expect(fixture.componentInstance.styles).toEqual([]);
        expect(JSON.stringify(fixture.componentInstance.icons)).toEqual(JSON.stringify([]));
    });

    it('subscribes to game classifier', () => {
        const fixture = TestBed.createComponent(GameMenuListComponent);

        let icons = new Map<string, string>();
        icons.set('C1', 'icon1');
        icons.set('Category 2', 'icon2');
        icons.set('Category Number.3', 'icon3');
        let categories = ['C1', 'Category 2', 'Category Number.3'];
        MockGameClassifier.classifications.next(categories);
        MockGameClassifier.icons.next(icons);
        fixture.detectChanges();

        expect(fixture.componentInstance.categories).toEqual(categories);
        let expectedStyles = ['c1', 'category-2', 'category-number3'];
        expect(fixture.componentInstance.styles).toEqual(expectedStyles);
        expect(JSON.stringify(fixture.componentInstance.icons)).toEqual(JSON.stringify(icons));

        let querySelectorAll = fixture.nativeElement.querySelectorAll('game-menu-category-list');
        expect(querySelectorAll.length).toEqual(3);
        expect(querySelectorAll[0].textContent).toEqual(categories[0] + expectedStyles[0] + 'icon1');
        expect(querySelectorAll[1].textContent).toEqual(categories[1] + expectedStyles[1] + 'icon2');
        expect(querySelectorAll[2].textContent).toEqual(categories[2] + expectedStyles[2] + 'icon3');

    });

    it('toggling help on/off shows/hides popover', fakeAsync(inject([HelpDisplayService], (helpService) => {
        const fixture = TestBed.createComponent(GameMenuListComponent);
        fixture.detectChanges();
        helpService.toggleHelp();
        tick();
        fixture.detectChanges();
        expect(fixture.nativeElement.querySelectorAll('.popover').length).toBeCloseTo(1);
        helpService.toggleHelp();
        tick();
        fixture.detectChanges();
        expect(fixture.nativeElement.querySelectorAll('.popover').length).toBeCloseTo(0);
    })));
});
