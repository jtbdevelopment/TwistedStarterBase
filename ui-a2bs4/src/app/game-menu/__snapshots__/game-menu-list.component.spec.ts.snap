// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:  game menu list component initializes 1`] = `
<app-game-menu-list
  categories={[Function Array]}
  gameClassifier={[Function MockGameClassifier]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  icons={[Function Map]}
  styles={[Function Array]}
>
  <div
    class="game-bar"
    ng-reflect-ngb-popover="Here are where your games are/"
    ng-reflect-placement="bottom"
    ng-reflect-popover-title="Games"
    ng-reflect-triggers="manual"
    ngbpopover="Here are where your games are/will be.  Click on one to play it."
    placement="bottom"
    popovertitle="Games"
    triggers="manual"
  >
    
      
    
    
  
  </div>
  
</app-game-menu-list>
`;

exports[`Component:  game menu list component subscribes to game classifier 1`] = `
<app-game-menu-list
  categories={[Function Array]}
  gameClassifier={[Function MockGameClassifier]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  icons={[Function Map]}
  styles={[Function Array]}
>
  <div
    class="game-bar"
    ng-reflect-ngb-popover="Here are where your games are/"
    ng-reflect-placement="bottom"
    ng-reflect-popover-title="Games"
    ng-reflect-triggers="manual"
    ngbpopover="Here are where your games are/will be.  Click on one to play it."
    placement="bottom"
    popovertitle="Games"
    triggers="manual"
  >
    
      
    
    <div
      class="list-group list-group-games list-group-c1"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="C1"
        ng-reflect-icon="icon1"
        ng-reflect-style="c1"
      >
        <p>
          C1
        </p>
        <p>
          c1
        </p>
        <p>
          icon1
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-2"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category 2"
        ng-reflect-icon="icon2"
        ng-reflect-style="category-2"
      >
        <p>
          Category 2
        </p>
        <p>
          category-2
        </p>
        <p>
          icon2
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-number3"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category Number.3"
        ng-reflect-icon="icon3"
        ng-reflect-style="category-number3"
      >
        <p>
          Category Number.3
        </p>
        <p>
          category-number3
        </p>
        <p>
          icon3
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    
  
  </div>
  
</app-game-menu-list>
`;

exports[`Component:  game menu list component toggling help on/off shows/hides popover 1`] = `
<app-game-menu-list
  categories={[Function Array]}
  gameClassifier={[Function MockGameClassifier]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  icons={[Function Map]}
  styles={[Function Array]}
>
  <div
    aria-describedby="ngb-popover-2"
    class="game-bar"
    ng-reflect-ngb-popover="Here are where your games are/"
    ng-reflect-placement="bottom"
    ng-reflect-popover-title="Games"
    ng-reflect-triggers="manual"
    ngbpopover="Here are where your games are/will be.  Click on one to play it."
    placement="bottom"
    popovertitle="Games"
    triggers="manual"
  >
    
      
    
    <div
      class="list-group list-group-games list-group-c1"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="C1"
        ng-reflect-icon="icon1"
        ng-reflect-style="c1"
      >
        <p>
          C1
        </p>
        <p>
          c1
        </p>
        <p>
          icon1
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-2"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category 2"
        ng-reflect-icon="icon2"
        ng-reflect-style="category-2"
      >
        <p>
          Category 2
        </p>
        <p>
          category-2
        </p>
        <p>
          icon2
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-number3"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category Number.3"
        ng-reflect-icon="icon3"
        ng-reflect-style="category-number3"
      >
        <p>
          Category Number.3
        </p>
        <p>
          category-number3
        </p>
        <p>
          icon3
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    
  
  </div><ngb-popover-window
    _nghost-c0=""
    class="popover bs-popover-bottom"
    id="ngb-popover-2"
    role="tooltip"
    style="top: 0px; left: 0px;"
  >
    
      
    <div
      _ngcontent-c0=""
      class="arrow"
    />
    
      
    <h3
      _ngcontent-c0=""
      class="popover-header"
    >
      Games
    </h3>
    <div
      _ngcontent-c0=""
      class="popover-body"
    >
      Here are where your games are/will be.  Click on one to play it.
    </div>
  </ngb-popover-window>
  
</app-game-menu-list>
`;

exports[`Component:  game menu list component toggling help on/off shows/hides popover 2`] = `
<app-game-menu-list
  categories={[Function Array]}
  gameClassifier={[Function MockGameClassifier]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  icons={[Function Map]}
  styles={[Function Array]}
>
  <div
    class="game-bar"
    ng-reflect-ngb-popover="Here are where your games are/"
    ng-reflect-placement="bottom"
    ng-reflect-popover-title="Games"
    ng-reflect-triggers="manual"
    ngbpopover="Here are where your games are/will be.  Click on one to play it."
    placement="bottom"
    popovertitle="Games"
    triggers="manual"
  >
    
      
    
    <div
      class="list-group list-group-games list-group-c1"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="C1"
        ng-reflect-icon="icon1"
        ng-reflect-style="c1"
      >
        <p>
          C1
        </p>
        <p>
          c1
        </p>
        <p>
          icon1
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-2"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category 2"
        ng-reflect-icon="icon2"
        ng-reflect-style="category-2"
      >
        <p>
          Category 2
        </p>
        <p>
          category-2
        </p>
        <p>
          icon2
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    <div
      class="list-group list-group-games list-group-category-number3"
    >
      
          
      <app-game-menu-category-list
        ng-reflect-category="Category Number.3"
        ng-reflect-icon="icon3"
        ng-reflect-style="category-number3"
      >
        <p>
          Category Number.3
        </p>
        <p>
          category-number3
        </p>
        <p>
          icon3
        </p>
      </app-game-menu-category-list>
      
      
    </div>
    
  
  </div>
  
</app-game-menu-list>
`;
